modules/misc/src/materials/ADDensity.C

```diff
@@ -26,12 +26,21 @@ ADDensity::validParams()
 ADDensity::ADDensity(const InputParameters & parameters)
   : ADMaterial(parameters),
     _coord_system(getBlockCoordSystem()),
-    _disp_r(adCoupledValue("displacements", 0)),
+    _disp_r(coupledComponents("displacements") ? adCoupledValue("displacements", 0) : _ad_zero),
     _initial_density(getParam<Real>("density")),
     _density(declareADProperty<Real>("density"))
 {
+  if (getParam<bool>("use_displaced_mesh"))
+    paramError("ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to "
+               "incorrect gradient values");
+
   // get coupled gradients
   const unsigned int ndisp = coupledComponents("displacements");
+
+  if (ndisp == 0 && _fe_problem.getDisplacedProblem())
+    mooseError(
```

Comment:
paramError

Reply:
Thought this one would not qualify as such. Addressed.